### PR TITLE
Add the zoned datetime data type to Memgraph

### PR DESCRIPTION
**Pull request by antepusic**
### Description

| Specification: [Zoned temporal types](https://www.notion.so/memgraph/Zoned-temporal-types-1d0de6c83b5a429180f28ec22d3e00fa)

This PR integrates the `utils::ZonedDateTime` object developed in #1850 in order to support zoned datetimes in Memgraph.
Additions to the query module APIs will be handled in a separate PR.

[master < Epic] PR
- [x] Write E2E tests
- [x] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [x] Provide the full content or a guide for the final git message
    > Add the zoned datetime data type to Memgraph

### Documentation checklist
- [x] Add the documentation label tag
- [x] Add the bug / feature label tag
- [x] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [x] Write a release note, including added/changed clauses
    > Added timezone support to Memgraph with the new zoned datetime type, which represents points in time that have a defined timezone. You can now work with temporal data from different locales, as well as easily migrate datasets with zoned values.
- [x] Link the documentation PR here
    - https://github.com/memgraph/documentation/pull/723
- [x] Tag someone from docs team in the comments

### CODE REVIEW

**Review by gvolfing — APPROVED** [review]
LGTM

File: src/storage/v2/property_store.cpp [hunk 1]
```diff
@@ -520,6 +591,69 @@ std::optional<uint64_t> DecodeTemporalDataSize(Reader &reader) {
...
+  }
+
+  auto type_value = reader.ReadUint(metadata->id_size);
+  if (!type_value) return std::nullopt;
+
+  auto microseconds_value = reader.ReadInt(metadata->payload_size);
+  if (!microseconds_value) return std::nullopt;
+
+  if (metadata->type == Type::ZONED_TEMPORAL_DATA) {
+    auto tz_str_length = reader.ReadUint(TZ_NAME_LENGTH_SIZE);
```
Comment (Darych) [inline src/storage/v2/property_store.cpp:609]: Maybe we can have a function in Reader class to read timezone in this format? The same as we have `ReadTimezoneOffset`

Comment (Darych) [reply src/storage/v2/property_store.cpp:609]: Or even better we can incapsulate logic of types in one function and return `utils::Timezone`.

File: tests/gql_behave/tests/memgraph_V1/features/match.feature [hunk 2]
```diff
@@ -469,6 +469,26 @@ Feature: Match
...
             | P13DT12H |
 
+    Scenario: Test match with order by and datetime
+        Given an empty graph
+        And having executed:
+            """
+            CREATE({a: DATETIME('2024-01-22T08:11:31[Etc/UTC]')}),
+                  ({a: DATETIME('2024-01-22T08:12:31[Etc/UTC]')}),
+                  ({a: DATETIME('2024-01-22T09:12:31[Europe/Zurich]')}),
+                  ({a: DATETIME('2024-01-22T09:12:31[Europe/Warsaw]')})
```
Comment (Darych) [inline tests/gql_behave/tests/memgraph_V1/features/match.feature:479]: Can we add edge cases like +18, -18 hours, not 00 minutes but [30, 45](https://www.timeanddate.com/time/time-zones-interesting.html).

File: tests/unit/query_expression_evaluator.cpp [hunk 3]
```diff
@@ -2386,4 +2454,38 @@ TYPED_TEST(FunctionTest, Duration) {
...
+              today.SysMicrosecondsSinceEpoch().count(), one_sec_in_microseconds);
+  EXPECT_EQ(this->EvaluateFunction("DATETIME", TypedValue(std::map<std::string, TypedValue>{})).ValueZonedDateTime(),
+            memgraph::utils::ZonedDateTime({{}, {}, memgraph::utils::DefaultTimezone()}));
+
+  EXPECT_THROW(this->EvaluateFunction("DATETIME", "{}"), memgraph::utils::BasicException);
+  EXPECT_THROW(
+      this->EvaluateFunction("DATETIME", TypedValue(std::map<std::string, TypedValue>{{"hours", TypedValue(1970)}})),
+      QueryRuntimeException);
+  EXPECT_THROW(
+      this->EvaluateFunction("DATETIME", TypedValue(std::map<std::string, TypedValue>{{"seconds", TypedValue(1970)}})),
```
Comment (Darych) [inline tests/unit/query_expression_evaluator.cpp:2488]: What will happen if you will pass just "year" or "month" or "day" or "millisecond" or "microsecond"?

File: src/communication/bolt/v1/decoder/decoder.hpp [hunk 4]
```diff
@@ -131,6 +131,18 @@ class Decoder {
             return ReadUnboundedEdge(data);
           case Signature::Path:
             return ReadPath(data);
+          case Signature::DateTime:
+            return ReadDateTime(data);
+          case Signature::DateTimeZoneId:
+            return ReadDateTimeZoneId(data);
+          case Signature::LegacyDateTime: {
+            if (major_v_ > 4) return false;
```
[the remark this answers is not in the document]
Comment (antepusic) [reply src/communication/bolt/v1/decoder/decoder.hpp:139]: 1. Per the [documentation](https://memgraph.com/docs/client-libraries) Memgraph does provide backward compatibility. In any case, we should support the pre-Bolt v5 structures because mgclient is v4.
2. The latest versions of mgconsole/Lab still open Bolt v4 connections, which means that Bolt uses the legacy structures for their values.
3. The legacy vs. latest distinction only applies to the Bolt protocol. In Memgraph, they are represented by the same type.
4. No (legacy datetime is a structure for zoned temporal values, and my PR is the one that will add support for them)

File: src/communication/bolt/v1/decoder/decoder.hpp [hunk 5]
```diff
@@ -630,5 +642,95 @@ class Decoder {
...
+  }
+
+  bool ReadDateTime(Value *data) {
+    int64_t time_since_epoch;
+    if (!ComputeTimeSinceEpoch(&time_since_epoch)) {
+      return false;
+    }
+
+    Value tz_offset_secs;
+    if (!ReadValue(&tz_offset_secs, Value::Type::Int)) {
```
[the remark this answers is not in the document]
Comment (antepusic) [reply src/communication/bolt/v1/decoder/decoder.hpp:673]: Yes. With zoned temporal values, the Bolt protocol specifies that each structure (`DateTime`, `DateTimeZoneId`, and `Legacy DateTime` & `Legacy DateTimeZoneId`) has a field with timezone data.

File: tests/unit/query_expression_evaluator.cpp (hunk 3 above)
Comment (antepusic) [reply tests/unit/query_expression_evaluator.cpp:2488]: In that case, the created zoned datetime value will use the default values for all other properties. I’ll add a test to verify that.

File: src/query/interpret/eval.hpp [hunk 6]
```diff
@@ -714,6 +752,9 @@ class ExpressionEvaluator : public ExpressionVisitor<TypedValue> {
         result.emplace("microsecond", TypedValue(lt.microsecond, ctx_->memory));
         return TypedValue(result, ctx_->memory);
       }
+      case TypedValue::Type::ZonedDateTime: {
+        throw QueryRuntimeException("Can't coerce `{}` to Map.", expression_result.ValueZonedDateTime().ToString());
```
[the remark this answers is not in the document]
Comment (antepusic) [reply src/query/interpret/eval.hpp:756]: Either way is doable; in this case I chose this way for the following reasons:
* It’s in line with other graph databases.
* The year...microsecond components aren’t actual members (whereas they are in e.g. `LocalDateTime`); they’re rather computed properties
* More consistent for users - If in the future we add new ZonedDateTime components (e.g. week, epochMicroseconds or offset), existing queries with AllPropertyLookup (i.e. map projections) will change since the map projection should contain them as new fields). By forbidding ZonedDateTime in AllPropertyLookup, this is avoided.

File: src/storage/v2/property_value.hpp [hunk 7]
```diff
@@ -290,8 +308,12 @@ inline std::ostream &operator<<(std::ostream &os, const PropertyValue &value) {
                            [](auto &stream, const auto &pair) { stream << pair.first << ": " << pair.second; });
       return os << "}";
     case PropertyValue::Type::TemporalData:
-      return os << fmt::format("type: {}, microseconds: {}", TemporalTypeTostring(value.ValueTemporalData().type),
+      return os << fmt::format("type: {}, microseconds: {}", TemporalTypeToString(value.ValueTemporalData().type),
                                value.ValueTemporalData().microseconds);
+    case PropertyValue::Type::ZonedTemporalData:
+      auto temp_value = value.ValueZonedTemporalData();
+      return os << fmt::format("type: {}, microseconds: {}, timezone: {}", ZonedTemporalTypeToString(temp_value.type),
+                               temp_value.IntMicroseconds(), temp_value.timezone.ToString());
```
[the remark this answers is not in the document]
Comment (antepusic) [reply src/storage/v2/property_value.hpp:316]: Either way works. I only didn’t make a dedicated function because this line would be the only place where it’s used.

File: src/storage/v2/property_store.cpp (hunk 1 above)
Comment (antepusic) [reply src/storage/v2/property_store.cpp:609]: Agree that encapsulation is a good idea, will do.

**Review by Darych — APPROVED** [review]

File: src/utils/temporal.hpp [hunk 8]
```diff
@@ -414,13 +464,33 @@ struct ZonedDateTime {
   friend ZonedDateTime operator-(const ZonedDateTime &zdt, const Duration &dur) { return zdt + (-dur); }
 
   friend Duration operator-(const ZonedDateTime &lhs, const ZonedDateTime &rhs) {
-    return Duration(lhs.MicrosecondsSinceEpoch()) - Duration(rhs.MicrosecondsSinceEpoch());
+    return Duration(lhs.SysMicrosecondsSinceEpoch().count()) - Duration(rhs.SysMicrosecondsSinceEpoch().count());
+  }
+
+  int64_t LocalYear() const { return int(LocalYMD().year()); }
+  uint64_t LocalMonth() const { return static_cast<unsigned>(LocalYMD().month()); }
+  uint64_t LocalDay() const { return static_cast<unsigned>(LocalYMD().day()); }
```
Comment (gvolfing) [inline src/utils/temporal.hpp:472]: Can we use static_cast<> instead of c-style cast on line 470?
also can we just static cast into the numbered type? e.g. `...return static_cast<uint64_t>(...`?

**Comment by antepusic** [comment]
@kgolubic Tagging you about the release note & docs

**Review by gvolfing — APPROVED** [review]

File: src/utils/temporal.hpp (hunk 8 above)
Comment (antepusic) [reply src/utils/temporal.hpp]: Fixed L470 to use `static_cast`. I use `static_cast<int>` and `static_cast<unsigned>` because other conversions aren’t available: `.year` returns `std::chrono::year`, which only has `operator int`. Analogously, `std::chrono::month` and `::day` only have `operator unsigned` available.
* https://en.cppreference.com/w/cpp/chrono/year
* https://en.cppreference.com/w/cpp/chrono/month
* https://en.cppreference.com/w/cpp/chrono/day
